vortex-array/src/array/varbin/stats.rs

```diff
@@ -15,10 +15,32 @@ impl ArrayStatisticsCompute for VarBinArray {
         if self.is_empty() {
             return Ok(StatsSet::new());
         }
+        if _stat == Stat::IsConstant {
+            return self.with_iterator(|iter| compute_is_constant(iter));
+        }
         self.with_iterator(|iter| compute_stats(iter, self.dtype()))
     }
 }
 
+pub fn compute_is_constant(iter: &mut dyn Iterator<Item = Option<&[u8]>>) -> StatsSet {
+    fn is_constant_stats_set(is_constant: bool) -> StatsSet {
+        StatsSet::from(HashMap::from([(Stat::IsConstant, is_constant.into())]))
+    }
+
+    match iter.next() {
```

Comment:
Itertools has the thing called `all_equal` which is a bit more succint. alternatively you can convert your for loop into `Iterator::all` which does the right thing and is a bit shorter

Reply:
Both `all_equal` and `all` want the `Iterator` to be `Sized`. I played a bit of type golf both seeing if we could switch to `impl Iterator ...` and if we could constrain the iterator to be `Sized` and I can't quite figure it out.

The simplest thing I tried re: `Sized` was:
```
error[E0225]: only auto traits can be used as additional traits in a trait object
  --> vortex-array/src/accessor.rs:12:73
   |
12 |         F: for<'a> FnOnce(&mut (dyn Iterator<Item = Option<&'a Item>> + Sized)) -> R;
   |                                     ---------------------------------   ^^^^^ additional non-auto trait
   |                                     |
   |                                     first non-auto trait
   |
   = help: consider creating a new trait with all of these as supertraits and using that trait here instead: `trait NewTrait: Iterator + Sized {}`
   = note: auto-traits like `Send` and `Sync` are traits that have special properties; for more information on them, visit <https://doc.rust-lang.org/reference/special-types-and-traits.html#auto-traits>
```
I also tried adding a `SizedIterator` but that seems to unearth different demons:
```
pub trait SizedIterator<T>: Iterator<Item = T> + Sized {}
```
```
error[E0038]: the trait `SizedIterator` cannot be made into an object
  --> vortex-array/src/array/bool/accessors.rs:18:62
   |
18 |               Validity::NonNullable | Validity::AllValid => Ok(f(&mut bools
   |  ______________________________________________________________^
19 | |                 .iter()
20 | |                 .map(|b| Some(if b { &TRUE } else { &FALSE })))),
   | |_______________________________________________________________^ `SizedIterator` cannot be made into an object
   |
note: for a trait to be "object safe" it needs to allow building a vtable to allow the call to be resolvable dynamically; for more information visit <https://doc.rust-lang.org/reference/items/traits.html#object-safety>
  --> vortex-array/src/accessor.rs:3:50
   |
3  | pub trait SizedIterator<T>: Iterator<Item = T> + Sized {}
   |           -------------                          ^^^^^ ...because it requires `Self: Sized`
   |           |
   |           this trait cannot be made into an object...
```

It vaguely seems to me that we might want something like:
```
pub trait ArrayAccessor {
    type Item: Sized;
    type Iterators;

    fn iterator(&self) -> &Iterators;
}
```
Which is implemented like:
```
enum BoolArrayIterator {
    AllValid(BitIterator) // or something like this?
    AllInvalid(NoneIterator) // again not sure on the element type here
    Mixed(Zip<...>)
}
impl ArrayAccessor<bool> for BoolArray {
    type Item = bool;
    type Iterators = BoolArrayIterators;
    fn iterator() {
        // match and cook up the right iterator
    }
}
```
This forces an explicit branch at the use site but everything is Sized and we can use itertools etc.